File: src/model/gpu_slot_model.cc
#include "gpu_slot_model.h"
#include "support.h"

GPU_JobSlot::GPU_JobSlot() {
	this->is_empty = true;
	this->sm = 0;
	this->sm_map = 0x0;
	this->job = nullptr;
	waitlist.reserve(20);
}

GPU_JobSlot::~GPU_JobSlot() {}

void GPU_JobSlot::assign_job_slot(Job *__job){
	this->job = __job;
	update_sm_map(__job->get_sm_map());
	is_empty = false;
}

void GPU_JobSlot::delete_job_slot() {
	this->job = nullptr;
	update_sm_map(0x0);
	is_empty = true;
	clear_waitlist();
}

void GPU_JobSlot::update_sm_map(uint64_t _sm_map) {
	this->sm_map = _sm_map;
	this->sm = get_sm_by_map(_sm_map);
}

/* Return the OCCUPIED number of SMs at tp */
const unsigned int GPU_JobSlot::get_sm_tp(float tp) const {
	if (!empty_()) {
		if (tp < this->job->gpu_end_time) {
			return this->job->get_sm();
		}
	}
	for (auto it : this->waitlist) {
		if (tp < it.gpu_end_time && tp >= it.gpu_start_time) {
			return it.get_sm();
		}
	}
	return 0;
}

const app_t GPU_JobSlot::get_app_tp(float tp) const {
	if (!empty_()) {
		if (tp < this->job->gpu_end_time) {
			return this->job->app_id;
		}
	}
	bool failed = true;
	for (auto it : this->waitlist) {
		if (tp < it.gpu_end_time && tp >= it.gpu_start_time) {
			return it.app_id;
		}
	}
	if (failed) {
		throw std::runtime_error("const app_t GPU_JobSlot::get_app_tp(float tp) const failed");
	}
}

File: src/model/gpu_slot_model.py
from typing import List

from src.model.job_model import Job

class GPU_JobSlot:
	def __init__(self, _sid):
		self.job: Job = None # the methods will not modify this feild
		self.is_empty: bool = True
		self.sm: int = 0 # occupied sm
		self.waitlist = []
		self.id = _sid

	def empty_(self) -> bool:
		return self.is_empty
	
	def empty_tp(self, tp: float) -> bool:
		if self.empty_() is False:
			if tp < self.job.gpu_end_time:
				return False
		for it in self.waitlist:
			if tp < it.gpu_end_time and tp >= it.gpu_start_time:
				return False
		return True
	
	def get_sm(self) -> int:
		return self.sm

	def get_sm_tp(self, tp: float) -> int:
		if self.empty_() is False:
			if tp < self.job.gpu_end_time:
				return self.job.get_sm()
		for it in self.waitlist:
			if tp < it.gpu_end_time and tp >= it.gpu_start_time:
				return it.get_sm()
		return 0

	def update_sm(self, _sm: int):
		self.sm = _sm
	
	def push_to_waitlist(self, _job: Job) -> None:
		self.waitlist.append(_job)
	
	def clear_waitlist(self) -> None:
		self.waitlist.clear()

	def assign_job_slot(self, _job: Job) -> None:
		self.job = _job
		self.update_sm(_job.get_sm())
		self.is_empty = False

	def delete_job_slot(self) -> None:
		self.job = None
		self.update_sm(0)
		self.is_empty = True
		self.clear_waitlist()

	def get_app_tp(self, tp: float) -> int:
		if self.empty_() is False:
			if tp < self.job.gpu_end_time:
				return self.job.app_id
		for it in self.waitlist:
			if tp < it.gpu_end_time and tp >= it.gpu_start_time:
				return it.app_id

	def calc_power_d(self, pd_ls: List[float], _job: Job) -> float:
		if _job == None:
			return 0
		return _job.get_sm() * pd_ls[_job.app_id]
